Implement iCloud GPS Accuracy Threshold

JustAmply opened the pull request:
Discards iCloud device tracker updates if the accuracy is larger than the configured threshold. The other sensors are not affected.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Wildly inaccurate location updates can often trigger things you don't want. Here's a fix for that.
The feature was supposedly implemented but the config-variables have stayed unused from what i could see. 
I just finished what was started a long time ago.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

This feature was supposed to be implemented all along, i think. Otherwise there wouldn't have been a (hidden) config-variable for it. So it could both be seen as a bugfix, as well as a new feature.

You could argue about the need of the option in the config-flow. If you want to keep it simple you could remove it, but the feature doesn't really add complexity and offers a real value, because if filters false-positive values.

- This PR fixes or closes issue: fixes #46261 
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Commit:
Implement GPS Accuracy Threshold

Discards iCloud device tracker updates if the accuracy is larger than the configurable threshold

homeassistant commented:
Hi @Justus502, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#47679).

Once you do that we will be able to review and accept this pull request.

Thanks!

Commit:
Add GPS Accuracy Threshold to config flow

Commit:
Get GPS Threshold from config

Commit:
Formatting with black

Commit:
Codeowners

Commit:
Fix protected-access for pylint

ntilley905 commented on homeassistant/components/icloud/account.py:
```diff
@@ -426,6 +436,18 @@ def update(self, status) -> None:
             ):
                 location = self._status[DEVICE_LOCATION]
+                if (
+                    self._config_entry.data[CONF_GPS_ACCURACY_THRESHOLD] is not None
```
No worries! Funnily enough, I learned about protected variables through a Home Assistant contribution code review. Go figure!

Justus502 replied:
I just changed it. Created a Getter-method in `IcloudAccount` instead of making the variable public. Honestly can't imagine why I didn't think of this before...

Commit:
Revert "Fix protected-access for pylint"

This reverts commit 1ee5ea5b37d41f35cecea0aff7cd1a587000b48e.

Commit:
Create Getter for Config Variable

Instead of getting from the config every time the device updates, just use the getter from the account

Commit:
Formatting

Commit:
Removing unused Import

Justus502 commented on homeassistant/components/icloud/account.py:
```diff
@@ -354,6 +354,11 @@ def fetch_interval(self) -> int:
         """Return the account fetch interval."""
         return self._fetch_interval
 
+    @property
+    def gps_accuracy_threshold(self) -> int:
+        """Return configured GPS Accuracy Threshold."""
+        return self._gps_accuracy_threshold
+
```
I'm not very familiar with Python but I'm really trying to get this stuff right, so bear with me. 

I just read online that it might not be good practice to use a getter/property in this context.
Should the variable `self._gps_accuracy_threshold` just be made public? So that `self.gps_accuracy_threshold` can be read in the device class?

Would appreciate some feedback so that the code-quality doesn't suffer :)

ntilley905 commented on homeassistant/components/icloud/account.py:
```diff
@@ -354,6 +354,11 @@ def fetch_interval(self) -> int:
         """Return the account fetch interval."""
         return self._fetch_interval
 
+    @property
```
I compared with the Google Maps device tracker, and in that component a public variable is used instead of a getter method. I think for consistency sake, we should use that here too.

Justus502 replied:
I was wondering what the best practice was here.
Will do that! Thanks!

ntilley905 commented on homeassistant/components/icloud/translations/de.json:
```diff
@@ -28,7 +28,8 @@
                     "password": "Passwort",
                     "username": "Email",
-                    "with_family": "Mit Familie"
+                    "with_family": "Mit Familie",
```
The translations will be done via Lokalize, don't do them here.

Justus502 replied:
Okay! Probably should have read up on the documentation a bit more, but it just seemed too tempting :)
Will remove it 👍

ntilley905 commented on homeassistant/components/icloud/translations/en.json:
```diff
@@ -29,7 +29,8 @@
                     "username": "Email",
-                    "with_family": "With family"
+                    "with_family": "With family",
+                    "gps_accuracy_threshold": "Discard inaccurate location updates threshold (meters)"
```
Make sure this is also added to `strings.json`.

Justus502 replied:
Done!

Commit:
Remove translation

Will be done with Localize

Commit:
Make variable public and remove property

like in the google_maps integration

Commit:
Improve wording of Config-flow string

Commit:
Update strings.json

Justus502 commented on homeassistant/components/icloud/translations/en.json:
```diff
@@ -29,7 +29,8 @@
                 "data": {
                     "password": "Password",
                     "username": "Email",
-                    "with_family": "With family"
+                    "with_family": "With family",
+                    "gps_accuracy_threshold": "Discard inaccurate location updates above threshold (meters)"
                 },
```
@ntilley905 Should this be removed as well? Will localise use the `strings.json` or this file?

ntilley905 replied:
Adding the text to `en.json` is fine, there just shouldn't be translations done via direct editing.

ntilley905 reviewed (approved):
LGTM!

emontnemery commented:
@Justus502 You've marked yourself as codeowner for the icloud integration. The purpose of that is to mark yourself as responsible for fixing bugs with icloud, not to mark that you've contributed code to the icloud integration. Can you confirm that is your intention?

emontnemery commented:
The implementation in the PR will discard any updates with poor accuracy.
That seems a bit too simple, don't we need to consider the current location as well, or is it guaranteed there will be additional updates with better accuracy?

Justus502 commented:
I agree that the way i tried to fix the issue maybe is not ideal and a bit hacky.
But i also had another issue that lead me to try the threshold approach rather than any other. Please hear me out.

If I opened the history graph (or logbook) of an iCloud device tracker there was a lot of _zone-jumping_ happening, even if the device wasn't close to those zones.
And i think that that's exactly because of the code-snippet you posted from the zone-component.
The `radius` of the location update is not even being considered when looking for a closer zone. Just the distance to the _center of a zone_ is important here.

https://github.com/home-assistant/core/blob/96c391af09d4cbbcd68ecc8374a3e0d445394d96/homeassistant/components/zone/__init__.py#L124

#### Illustration of other problem
A device is physically inside zone A but sends out a new location update that's inaccurate and covers a huge area of multiple zones. The zone-integration thinks that the device is inside of zone B, just because zone B is supposedly closer to those wildly inaccurate GPS-coordinates of the device.
The accuracy should be considered here, I think. But that would entail adding a new threshold to the zone-integration.

#### Other device-tracker integrations use thresholds
Zone-jumping and highly inaccurate location updates happen pretty much just with the iCloud-integration in my experience!
The official HA-App and all other trackers i tried each seem to posses a mechanism that blocks highly inaccurate updates. They are either not even being pushed to HA or they are being discarded by the corresponding HA-integration.

#### Conclusion
I agree that the best solution would indeed be, to remove the GPS threshold for this integration. But that would only make sense if all device-tracker accuracy thresholds would be removed in the long run, just to add back appropriate thresholds in the integrations that use this data, like zone or the lovelace-map with timeline.
Is this planned? Because if not, then I don't think that this PR is half-bad.

frenck commented:
This PR isn't fixing zone jumping, as zone calculations take GPS inaccuracy into account, see:

https://github.com/home-assistant/core/blob/96c391af09d4cbbcd68ecc8374a3e0d445394d96/homeassistant/components/zone/__init__.py#L123

Displaying/consuming data is a different purpose. This PR changes in the input, while those the problem described is output (e.g., map showing in the frontend). Filtering the input changes/breaks other behavior. It's up to the consumer of the data to make an interpretation and e.g., filter out outliers.

emontnemery commented:
@Justus502 We don't solve these problems by simply discarding location state updates with accuracy worse than a certain threshold. The _consumers_ of location states need to be smarter and do filtering as needed.
For zones, enter and exit criteria for zones need to be smart, for example: don't enter a zone which is smaller than the location's accuracy, don't exit a zone if the new location is still in the zone considering accuracy.

In the map display we just present the raw data. We could amend that with drawing circles to represent the accuracy, or filter out inaccurate state updates if there are more accurate states available. Here's a blog post that may have some useful ideas: https://medium.com/@mizutori/make-it-even-better-than-nike-how-to-filter-locations-tracking-highly-accurate-location-in-1c9d53d31d93

In both cases, PRs would be welcome to improve the consumption of location states.

Justus502 commented:
> This PR isn't fixing zone jumping, as zone calculations take GPS inaccuracy into account, see:
> 
> https://github.com/home-assistant/core/blob/96c391af09d4cbbcd68ecc8374a3e0d445394d96/homeassistant/components/zone/__init__.py#L123
> 
> In that line, `radius` is the GPS inaccuracy passed in by the iCloud integration.

I have a problem with that statement. Because it doesn't take wildly inaccurate GPS updates into account.

To illustrate further I prepared a little image:
<img src="https://user-images.githubusercontent.com/15849513/113349655-acb61d80-9338-11eb-9796-b8a566261d32.png" width="500" height="500">
The **orange** circles are the GPS-coordinates in the center and the (in)accuracy sent by the device.
The **green** ones are zones and their center in the middle.
The **red** circle is the _actual location_ of the device.

To calculate the active zone, first all zones closer than the GPS inaccuracy are calculated. So all green zones shown here are `within_zone = true`. But when looking for the closest zone, just the distance between the given GPS-data and the middle of the zones is calculated.
And here lies the problem in my eyes. The distance from the GPS-coordinates to the middle of the upper-left zone is smaller than any to other, so the state of the device-tracker is getting set to that zone.
There is no problem doing that if the inaccuracy is low enough for it not to matter and can even be desirable, but with inaccuracies i encountered of over 5000 meters this becomes a problem.
This may be a design decision in the zone-integration, but I don't want false positives or zone-jumping if they rely on such inaccurate data.

With other device-trackers this is no problem because their input is most likely already filtered somehow.

> Displaying/consuming data is a different purpose. This PR changes in the input, while those the problem described is output (e.g., map showing in the frontend). Filtering the input changes/breaks other behavior. It's up to the consumer of the data to make an interpretation and e.g., filter out outliers.

I agree with you. Changing the input is hacky and not best practice.
Another solution would be better.

Justus502 commented:
> @Justus502 We don't solve these problems by simply discarding location state updates with accuracy worse than a certain threshold. The _consumers_ of location states need to be smarter and do filtering as needed.

Okay, I see your point and 100% agree. But on the other hand other integrations are doing exactly that from what i could find out.

> For zones, enter and exit criteria for zones need to be smart, for example: don't enter a zone which is smaller than the location's accuracy, don't exit a zone if the new location is still in the zone considering accuracy.

That would be a great fix! This currently isn't implemented, is it?

> In the map display we just present the raw data. We could amend that with drawing circles to represent the accuracy, or filter out inaccurate state updates if there are more accurate states available. Here's a blog post that may have some useful ideas: https://medium.com/@mizutori/make-it-even-better-than-nike-how-to-filter-locations-tracking-highly-accurate-location-in-1c9d53d31d93

This would be a good change as well, i agree!
Thank you btw! Very interesting article!

emontnemery commented:
@Justus502 PR(s) improving how the position states are consumed by the zone and map would be most welcome.

emontnemery commented:
@Justus502 Your illustration was extremely helpful 👍 
Here's a draft PR attempting to improve the logic for entering and leaving zones: https://github.com/home-assistant/core/pull/48860
